### ISSUE TITLE
Serialization of DatabaseChangeLog doesn't produce correct result with 'preConditions'

### ISSUE THREAD

**Issue by RadekWikturna** [liquibase/liquibase#4379]
### Search first

- [X] I searched and no similar issues were found

### Description

Background: in our project (SpringBoot, Kotlin) we use Liquibase not only to run database migration for the project itself, but also for custom applications which are manually customized as a set of configuration files. 
Entities are described in a json file - a simplified version might look like this:

```
{
  name: Team
  fields: [
    {
      name: name
      type: String
      length(40)
    }
...
 ]
}
```

Based on this json file, we generate a database changelog (instance of `liquibase.changelog.DatabaseChangeLog`) programmatically using Liquibase API from `liquibase.change.core` package (classes such as `CreateTableChange`, `ColumnConfig`, `AddColumnChange`...)
and finally we serialize the changelog to String (and store it to a file) using `YamlChangeLogSerializer`.

simplified code:

```
fun serializeDbChangeLog(databaseChangelog: DatabaseChangeLog): String {
    val changeLogSerializer = YamlChangeLogSerializer()
    val outputStream = ByteArrayOutputStream()
    changeLogSerializer.write(changeSets, outputStream)
    return outputStream.toString(Charsets.UTF_8)
}
```

The generated output for the entity might look like this:

```
databaseChangeLog:
- changeSet:
    id: '1'
    author: test
    objectQuotingStrategy: LEGACY
    changes:
    - createTable:
        columns:
        - column:
            constraints:
              nullable: false
              primaryKey: true
            name: sys_pk
            type: INT
        - column:
            constraints:
              nullable: false
            name: name
            type: VARCHAR(40)
        tableName: team
```

Everything works nicely, except when we needed pro programatically generate 'preConditions' using some classes in the `liquibase.precondition` package. The expected serialized form was for example:

```
databaseChangeLog:
- changeSet:
    id: '1'
    author: test
    objectQuotingStrategy: LEGACY
    preConditions:
    - or:
      - sqlCheck:
          expectedResult: 0
          sql: select count(*) from team
      - not:
        - sqlCheck:
            expectedResult: 0
            sql: select count(*) from referenced_table
      onError: HALT
      onFail: HALT
      onFailMessage: Some message
      onSqlOutput: IGNORE
```

but when serialized, it looks like this (and this is unparseable):

```
databaseChangeLog:
- changeSet:
    id: '1'
    author: test
    objectQuotingStrategy: LEGACY
    preconditions:
      preConditions:
        nestedPreconditions:
        - or:
            nestedPreconditions:
            - sqlCheck:
                expectedResult: '0'
                sql: select count(*) from team
            - not:
                nestedPreconditions:
                - sqlCheck:
                    expectedResult: '0'
                    sql: select count(*) from referenced_table
        onError: !!liquibase.precondition.core.PreconditionContainer$ErrorOption 'HALT'
        onFail: !!liquibase.precondition.core.PreconditionContainer$FailOption 'HALT'
        onFailMessage: Some message
        onSqlOutput: !!liquibase.precondition.core.PreconditionContainer$OnSqlOutputOption 'IGNORE'
```

As you can see, this contains not only extra 'preConditions' node, but also 'nestedPreconditions'. And the enum values are serialized wit full class name.

I though I was doing something wrong so I worte the the changelog with 'preConditions' manually, had it parsed by `YamlChangeLogParser()` and serialized again by `YamlChangeLogSerializer` bit the result was the same - unparseable changelog.

### Steps To Reproduce

I wrote a simple SpringBoot/Kotlin/Liquibase project that tests some found Liquibase bugs. It has one test class `LiquibaseBugsApplicationTests`. For this test, its enough to run the test `parse and serialize db changelog with preconditions produces the same result`. When you run it, it parses and serializes a changelog and compares whether it is the same - it fails.

[LiquibaseBugs.zip](https://github.com/liquibase/liquibase/files/11756994/LiquibaseBugs.zip)

### Expected/Desired Behavior

Programmatically created `DatabaseChangeLog` containing `preConditions` can be correctly serialized by `YamlChangeLogSerializer` and should be parseable  by `YamlChangeLogParser`

### Liquibase Version

4.20

### Database Vendor & Version

H2, mySQL

### Liquibase Integration

SpringBoot 3

### OS and/or Infrastructure Type/Provider

Windows 10

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR (Thank you!)

**Comment by kevin-atx** [comment]
@filipelautert - is this a pervasive issue that impacts serialization for all changelog types, or does it only affect JSON output? Could you provide guidance in case @RadekWikturna or other SpringBoot users want to address this issue?

**Comment by filipelautert** [comment]
@kevin-atx this seems to be related to all kind of serializations. It may be related to the way that `CustomPreconditionWrapper` methods `getSerializableFields` and `getSerializableFieldValue` are configured, and the problem seems to be here since the beginning.  It will need some debugging to understand what is happening under the hood and to fix it.

**Comment by rH4rtinger** [comment]
What is the current status of this issue?

**Comment by MalloD12** [comment]
Hi @RadekWikturna,

I have made some code changes related to this issue that is producing the below output:

```code:yaml
databaseChangeLog:
- changeSet:
    id: '1'
    author: test
    objectQuotingStrategy: LEGACY
    preconditions:
      preConditions:
        nestedPreconditions:
        - or:
            nestedPreconditions:
            - sqlCheck:
                expectedResult: '0'
                sql: select count(*) from team
            - not:
                nestedPreconditions:
                - sqlCheck:
                    expectedResult: '0'
                    sql: select count(*) from referenced_table
        onError: HALT
        onFail: HALT
        onFailMessage: Some message
        onSqlOutput: IGNORE
```

I know this is still containing the properties (`preConditions` and `nestedPreconditions`) you mentioned are wrong but we think that is the correct behavior as those properties are part of the related objects, so we think that is correct they are returned by the `YamlSerializer`. What we think is not correct (up to date) are precondition [docs](https://docs.liquibase.com/concepts/changelogs/preconditions.html).

**Comment by MalloD12** [comment]
@adrian-velonis1 - @AMBERMW13:

Docs says in their examples preconditions should be specified for example as:

```
- preConditions:
    - or:
        - CONDITION_1: VALUE
        - CONDITION_2:
           SUB_CONDITION: SUB_VALUE
```

but it should be like:
```
preconditions:
      preConditions:
        nestedPreconditions:
        - or:
              nestedPreconditions:
 ```

Since as we mentioned they are part of the preconditions-related objects (`PreconditionsContainer`, `PreconditionLogic`, etc).

Thanks,
Daniel.

**Comment by adrian-velonis1** [comment]
[DOC INTERNAL] https://datical.atlassian.net/browse/PD-4954

**Comment by RadekWikturna** [comment]
I've just now noticed that the issue has been closed. (sorry, my Github mails go to a different folder...)

I've re-ran my test case with Liquibase 4.30 and a generated change set with preconditions is now serialized without the class names for enums and it can be parsed. This is good news.

However, it's surprising to hear that there are 2 preconditions formats and that the documentation should even document the new one...I think it would only confuse the reader. The fact that a programmatically generated preconditions are serialized in different way (with several nested nodes) is an **implementation detail that leaks to the API** for the user and if you now document it, you will never be able to get rid of it as it will become part of the API.
